### PR TITLE
DISPATCH-1566: fix safe_snptrintf

### PR DESCRIPTION
**Pull request by nicob87**
As it is explained in [DISPATCH-1566](https://issues.apache.org/jira/browse/DISPATCH-1566):
"This function fails for size = 0, and ... inside it calls vsnprintf, without considering that in case of error vsnprintf returns a negative number."

So after fix we can expect this behaviour:
- (size = 0) -> safe_snprintf returns 0, and does nothing. 
- (size > INT_MAX) -> safe_snprintf returns 0, and does nothing. Since inner vsnprintf function returns int, it has no sense (and it may produce an error) to accept size > INT_MAX.
- (inner vsnprintf return value > (size -1)) -> safe_snprintf returns (size-1) -> which is the max possible value.
- (inner vsnprintf returns negative number) -> safe_snprintf returns 0. there was a parsing error.
- (inner vsnprintf return value >= 0 and < (size-1)) -> safe_snprintf returns vsnprintf.

This behaviour has been unit-tested but those changes will be included in other pr that is blocked by: 
[DISPATCH-1568](https://issues.apache.org/jira/browse/DISPATCH-1568)

### CODE REVIEW

**Comment by nicob87** [comment]
Test error is only in the RelWithDebInfo build, and it is in the tear down, If I am not wrong it is not related to this pr.

```
51: ======================================================================
51: ERROR: tearDownClass (system_tests_edge_router.LinkRouteProxyTest)
51: ----------------------------------------------------------------------
51: Traceback (most recent call last):
51:   File "/home/travis/build/apache/qpid-dispatch/tests/system_test.py", line 691, in tearDownClass
51:     cls.tester.teardown()
51:   File "/home/travis/build/apache/qpid-dispatch/tests/system_test.py", line 637, in teardown
51:     raise RuntimeError("Errors during teardown: \n\n%s" % "\n\n".join([str(e) for e in errors]))
51: RuntimeError: Errors during teardown: 
51: 
51: Process 14088 error: exit code 23, expected 0
51: qdrouterd -c EB1.conf -I /home/travis/build/apache/qpid-dispatch/python
51: /home/travis/build/apache/qpid-dispatch/build/tests/system_test.dir/system_tests_edge_router/LinkRouteProxyTest/setUpClass/EB1-8.cmd
51: >>>>
```

File: src/router_core/terminus.c [hunk 1]
```diff
@@ -75,24 +76,34 @@ void qdr_terminus_free(qdr_terminus_t *term)
...
     int rc = vsnprintf(str, size, format, ap);
     va_end(ap);
 
-    if (size && rc >= size)
-        return size - 1;  // return actual # of bytes written (excluding null)
-    return rc;
-}
+    if (rc < 0) { // parsing error!
+        //TODO log a warning somewhere?
+        return 0;
```
[the remark this answers is not in the document]
Comment (nicob87) [reply src/router_core/terminus.c]: Sharp eyes! ok I will do that change, and also I will add a unit-test for that condition (in the unit-tests pr)

[the remark this answers is not in the document]
Comment (nicob87) [reply src/router_core/terminus.c]: Done!
JFYI: the last commit in this WIP pr contains the this fix, and also the unittests (which passes).
https://github.com/apache/qpid-dispatch/pull/684

File: src/router_core/terminus.c [hunk 2]
```diff
@@ -75,24 +76,37 @@ void qdr_terminus_free(qdr_terminus_t *term)
...
 // wrapper will never return >= size, even if truncated.  This makes it safe to
 // do pointer & length arithmetic without overflowing the destination buffer in
 // qdr_terminus_format()
 //
-static inline int safe_snprintf(char *str, size_t size, const char *format, ...)
-{
+size_t safe_snprintf(char *str, size_t size, const char *format, ...) {
+    // max size allowed must be INT_MAX (since vsnprintf reutrns an int)
+    if (size == 0 || size > INT_MAX) {
+        //TODO log a warning somewhere?
```
Comment (ganeshmurthy) [inline src/router_core/terminus.c]: Please remove these TODOs. I don't think we are going to log anything here.

Comment (nicob87) [reply src/router_core/terminus.c]: Done

File: src/router_core/terminus.c [hunk 3]
```diff
@@ -75,24 +76,34 @@ void qdr_terminus_free(qdr_terminus_t *term)
...
-
 // DISPATCH-1461: snprintf() is evil - it returns >= size on overflow.  This
 // wrapper will never return >= size, even if truncated.  This makes it safe to
 // do pointer & length arithmetic without overflowing the destination buffer in
 // qdr_terminus_format()
 //
-static inline int safe_snprintf(char *str, size_t size, const char *format, ...)
-{
+static inline size_t safe_snprintf(char *str, size_t size, const char *format, ...) {
+    // max size allowed must be INT_MAX (since vsnprintf reutrns an int)
```
Comment (jdanekrh) [inline src/router_core/terminus.c]:
```suggestion
    // max size allowed must be INT_MAX (since vsnprintf returns an int)
```

Comment (nicob87) [reply src/router_core/terminus.c]: Fixed, Thanks Jiri!

**Pull request merged**